### PR TITLE
1.7 Local Persistent Volume design doc

### PR DESCRIPTION
**Pull request by msau42**
Migrating the design doc in google docs to Markdown, and updating it to reflect the implementation we did for 1.7.

Same sections:
* Background
* Use Cases
* Environments
* Goals/non-goals
* Plugin design
* Initial volume config

Updated sections:
* Plugin API changes
* Provisioner
* Test Cases
* Feature Plan

New sections:
* Node affinity
* Provisioner boostrapper
* 1.7 Limitations
* 1.9 Outline

### CODE REVIEW

File: contributors/design-proposals/local-storage-pv.md [hunk 1]
```diff
@@ -0,0 +1,631 @@
...
+
+The name of the PV needs to be unique across the cluster.  The provisioner will hash the node name,
+StorageClass name, and base file name in the volume path to generate a unique name.
+
+##### Packaging
+
+The provisioner is packaged as a container image and will run on each node in the cluster as part of
+a DaemonSet.  It needs to be run with a user or service account with the following permissions:
+
+* Create/delete/list/get PersistentVolumes - Can use the `system:persistentvolumeprovisioner` ClusterRoleBinding
```
Comment (liggitt) [inline contributors/design-proposals/local-storage-pv.md]: This gives every node write access to all PVs, which is not intended

File: contributors/design-proposals/local-storage-pv.md [hunk 2]
```diff
@@ -0,0 +1,631 @@
...
+
+##### API Changes
+
+For the initial alpha phase, node affinity is expressed as an optional
+annotation in the PersistentVolume object.
+
+```
+// AlphaStorageNodeAffinityAnnotation defines node affinity policies for a PersistentVolume.
+// Value is a string of the json representation of type NodeAffinity
+AlphaStorageNodeAffinityAnnotation = "volume.alpha.kubernetes.io/node-affinity"
```
Comment (liggitt) [inline contributors/design-proposals/local-storage-pv.md]: is the plan to migrate the annotation to a field on the PV spec or drop the alpha annotation when the field is added?

**Comment by zhouhaibing089** [comment]
cc @chengyli @wenlxie

File: contributors/design-proposals/local-storage-pv.md [hunk 3]
```diff
@@ -0,0 +1,631 @@
...
+Local PVs are statically created and not dynamically provisioned for the first phase.
+To mitigate the amount of time an administrator has to spend managing Local volumes,
+a Local static provisioner application will be provided to handle common scenarios.  For
+uncommon scenarios, a specialized provisioner can be written.
+
+The Local static provisioner will be developed in the
+[kubernetes-incubator/external-storage](https://github.com/kubernetes-incubator)
+repository, and will loosely follow the external provisioner design, with a few differences:
+
+* A provisioner instance needs to run on each node and only manage the local storage on its node.
```
[the remark this answers is not in the document]
Comment (msau42) [reply contributors/design-proposals/local-storage-pv.md]: The permissions are handled through ClusterRoleBinding.  But I am unaware of a way to have a different ClusterRoleBinding for each instance of a DaemonSet so that it only can see the PVs for that node.

Each provisioner instance will filter the PVs it handles in its code, but the PV informer itself is not filtered.  Each instance gives itself a name of "local-provisioner-<node uid>", and marks every PV It creates with that name in the "provisioned-by" annotation (part of the general [external provisioner](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/volume-provisioning.md)  design).  And it filters the PVs on that annotation.

If we want to filter at the informer level, I could also label the PVs with the provisioner name and filter on labels.  But that's still just relying on the implementation to correctly filter itself.  I would be open to any suggestions you may have.

File: contributors/design-proposals/local-storage-pv.md (hunk 1 above)
Comment (msau42) [reply contributors/design-proposals/local-storage-pv.md]: This is true of any external-provisioner in general.  The external provisioner design relies on the provisioner implementation to filter the PVs and only operate on PVs that it created by marking and checking the "provisioned-by" annotation.  Perhaps that design needs to be revisited so that you could create a rolebinding on a reduced set.

Comment (liggitt) [reply contributors/design-proposals/local-storage-pv.md]: Correct, but typically an external provisioner would not be required to run on all nodes, so you could avoid running it on a node containing untrusted workloads

File: contributors/design-proposals/local-storage-pv.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (msau42) [reply contributors/design-proposals/local-storage-pv.md]: Corrected the above comment to add nodeUID to the name.

File: contributors/design-proposals/local-storage-pv.md [hunk 4]
```diff
@@ -0,0 +1,631 @@
...
+* Allow pods to access local volumes without needing to undestand the storage
+layout on every node.
+* Persist local volumes and provide data gravity for pods.  Any pod
+using the local volume will be scheduled to the same node that the local volume
+is on.
+* Allow tolerations on volumes to be specified to release the volume during
+failure conditions, such as node, storage or scheduling failures.
+* Allow pods to specify local storage as part of a Deployment or StatefulSet.
+* Allow Administrators to set up and configure local volumes with simple methods.
+* Do not require qdministrators to manage the local volumes once provisioned
```
Comment (xiang90) [inline contributors/design-proposals/local-storage-pv.md]: qdministrators -> administrators

File: contributors/design-proposals/local-storage-pv.md [hunk 5]
```diff
@@ -0,0 +1,631 @@
...
+* Allow pods to mount any local block or filesystem based volume.
+* Allow pods to mount dedicated local disks, or channeled partitions as volumes for
+IOPS isolation.
+* Allow pods do access local volumes without root privileges.
+* Allow pods to access local volumes without needing to undestand the storage
+layout on every node.
+* Persist local volumes and provide data gravity for pods.  Any pod
+using the local volume will be scheduled to the same node that the local volume
+is on.
+* Allow tolerations on volumes to be specified to release the volume during
```
Comment (xiang90) [inline contributors/design-proposals/local-storage-pv.md]: not sure i understand this. can we explain this more clear?

File: contributors/design-proposals/local-storage-pv.md [hunk 6]
```diff
@@ -0,0 +1,631 @@
...
+persistent volumes where the data can persist beyond the lifetime of the pod.
+
+There is increasing demand for using local storage with persistent volumes,
+especially for distributed filesystems and databases such as GlusterFS and
+Cassandra.  The main motivations for using local storage include:
+
+* Performance:  Local SSDs achieve higher IOPS and throughput than a cloud
+provider’s remote storage.
+
+* Cost: High-performance networked storage is expensive to setup and maintain.
```
Comment (xiang90) [inline contributors/design-proposals/local-storage-pv.md]: if the upper layer application already does replication (and is not file/block level replication aware), it usually makes sense to just use local disk and let the application take over replication instead of using replicated block device.

File: contributors/design-proposals/local-storage-pv.md [hunk 7]
```diff
@@ -0,0 +1,631 @@
...
+* Allow pods do access local volumes without root privileges.
+* Allow pods to access local volumes without needing to undestand the storage
+layout on every node.
+* Persist local volumes and provide data gravity for pods.  Any pod
+using the local volume will be scheduled to the same node that the local volume
+is on.
+* Allow tolerations on volumes to be specified to release the volume during
+failure conditions, such as node, storage or scheduling failures.
+* Allow pods to specify local storage as part of a Deployment or StatefulSet.
+* Allow Administrators to set up and configure local volumes with simple methods.
```
Comment (zhangxiaoyu-zidif) [inline contributors/design-proposals/local-storage-pv.md]: Administrators  -> administrators

File: contributors/design-proposals/local-storage-pv.md (hunk 5 above)
Comment (msau42) [reply contributors/design-proposals/local-storage-pv.md]: The default behavior with local storage is that if the node goes down, your pod cannot be rescheduled because you're bound to that local volume.  If your application uses replication, it can potentially be rescheduled, on a brand new node, with a new volume and then resync the replica.  So the idea of tolerations here is similar to the pod tolerations feature, except at a volume level.  The tolerations setting will determine how long we wait on specific error conditions (like node unavailable) before we unbind the PVC to find a new one.

File: contributors/design-proposals/local-storage-pv.md (hunk 6 above)
Comment (msau42) [reply contributors/design-proposals/local-storage-pv.md]: Good point.  I think I mentioned it already in the distributed db and fs use case, but will add it here too.

File: contributors/design-proposals/local-storage-pv.md [hunk 8]
```diff
@@ -0,0 +1,631 @@
...
+
+### Phase 2
+
+* Block devices and raw partitions
+* Smarter PV binding to consider local storage and pod scheduling constraints
+
+### Phase 3
+
+* Support common partitioning patterns
+* Volume forgiveness
```
Comment (xiang90) [inline contributors/design-proposals/local-storage-pv.md]: what is this? needs more details or link to another doc?

Comment (msau42) [reply contributors/design-proposals/local-storage-pv.md]: This is the tolerations idea I mentioned above.  I'm leaving the design details to a future release when we actually have time to work on it.  There's too many features needed for local volumes to design it all in one go.

File: contributors/design-proposals/local-storage-pv.md [hunk 9]
```diff
@@ -0,0 +1,631 @@
...
+In the above example, the generated `mountDir` is `/mnt/local-storage/mnt ~ssds` and
+`/mnt/local-storage/mnt~hdds`, respectively.
+
+
+#### Limitations
+
+This initial phase for Local PV support will provide the following:
+
+* Local directories to be specified as Local PVs with node affinity
+* Pod using a PVC that is bound to a Local PV will always be scheduled to that node
```
Comment (xiang90) [inline contributors/design-proposals/local-storage-pv.md]: what is the alternatives? why this is an limitation?

File: contributors/design-proposals/local-storage-pv.md (hunk 5 above)
Comment (xiang90) [reply contributors/design-proposals/local-storage-pv.md]: ok. makes sense. basically pod just wants a local pv, but can forgive its state. maybe we can make this clear in the doc.

File: contributors/design-proposals/local-storage-pv.md (hunk 8 above)
Comment (xiang90) [reply contributors/design-proposals/local-storage-pv.md]: thanks.

File: contributors/design-proposals/local-storage-pv.md (hunk 9 above)
Comment (msau42) [reply contributors/design-proposals/local-storage-pv.md]: Sorry i need to organize this section better.  I wanted to start off with the use cases that are supported in 1.7, and then follow up with the limitations in 1.7.  So this first section is describing what a user can do.

File: contributors/design-proposals/local-storage-pv.md (hunk 1 above)
Comment (msau42) [reply contributors/design-proposals/local-storage-pv.md]: What about something like this: Have a single master provisioner that manages and operates on the PV objects.  Have it talk to worker pods on each node.  Each worker pod is only allowed to:

1. Report what local volumes it discovered.
2. Handle cleanup of its local volumes when requested by the master.
3. In the future, provision a new volume from a shared pool when requested by the master.

This way all the PV objects are managed by a single pod that can run on a trusted node, and each worker pod can only touch the local volumes on it's node.  The worker pods don't even need to look at or operate on PV objects.

**Comment by xiang90** [comment]
@msau42 Thanks for the doc! Looks pretty solid to me in general.

File: contributors/design-proposals/local-storage-pv.md [hunk 10]
```diff
@@ -0,0 +1,632 @@
...
+dedicated disks; others may only need small partitions and are ok with sharing disks.
+Instead of forcing a partitioning scheme on storage administrators, the Local volume
+is represented by a path, and lets the administrators partition their storage however they
+like, with a few minimum requirements:
+
+* The paths to the mount points are always consistent, even across reboots or when storage
+is added or removed.
+* The paths are backed by a filesystem (block devices or raw partitions are not supported for
+the first phase)
+* The storage is mounted according to the provisioner discovery scheme defined below
```
Comment (ddysher) [inline contributors/design-proposals/local-storage-pv.md]: Is this a minimal requirement?  My understanding is that admin can use local volume without provisioner so doesn't have to conform to provisioner discovery scheme.

File: contributors/design-proposals/local-storage-pv.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (ddysher) [reply contributors/design-proposals/local-storage-pv.md]: Is it possible to extend [node authorization](https://kubernetes.io/docs/admin/authorization/node/)?

File: contributors/design-proposals/local-storage-pv.md [hunk 11]
```diff
@@ -0,0 +1,632 @@
...
+To mitigate the amount of time an administrator has to spend managing Local volumes,
+a Local static provisioner application will be provided to handle common scenarios.  For
+uncommon scenarios, a specialized provisioner can be written.
+
+The Local static provisioner will be developed in the
+[kubernetes-incubator/external-storage](https://github.com/kubernetes-incubator)
+repository, and will loosely follow the external provisioner design, with a few differences:
+
+* A provisioner instance needs to run on each node and only manage the local storage on its node.
+* For this phase, it does not handle dynamic provisioning.  Instead, it performs static provisioning
```
Comment (ddysher) [inline contributors/design-proposals/local-storage-pv.md]: s/this phase/initial phase

The doc will keep updating, so `this phase` can be ambiguous :)

File: contributors/design-proposals/local-storage-pv.md [hunk 12]
```diff
@@ -0,0 +1,632 @@
...
+
+PV deletion will operate on the Update event.  If the PV it provisioned changes to the “Released”
+state, and if the reclaim policy is Delete, then it will cleanup the volume and then delete the PV,
+removing it from the cache.
+
+PV creation does not operate on any informer events.  Instead, it periodically monitors the discovery
+directories, and will create a new PV for each path in the directory that is not in the PV cache.  It
+sets the "pv.kubernetes.io/provisioned-by" annotation so that it can distinguish which PVs it created.
+
+For this phase, the allowed discovery file types are directories and mount points.  The PV capacity
```
Comment (ddysher) [inline contributors/design-proposals/local-storage-pv.md]: ditto

File: contributors/design-proposals/local-storage-pv.md (hunk 1 above)
Comment (liggitt) [reply contributors/design-proposals/local-storage-pv.md]: central management of PV resources coordinated with local management of local volumes, where the central manager only pays attention to what the local worker reports for volumes that are actually bound to the node seems better

File: contributors/design-proposals/local-storage-pv.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (liggitt) [reply contributors/design-proposals/local-storage-pv.md]: to do that, several things would be required:
* the instance on each node would have to run with the node's credentials (not as a daemonset, etc)
* there would have to be strong ties from a PV to a particular node (not just a label selector or annotation)

at first glance, a division of responsibilities like https://github.com/kubernetes/community/pull/989#discussion_r135700184 seems like it might work better

File: contributors/design-proposals/local-storage-pv.md [hunk 13]
```diff
@@ -0,0 +1,632 @@
...
+for a node.
+
+## Non-Goals
+
+* Provide data availability for a local volume beyond its local node.
+* Support the use of HostPath volumes and Local PVs on the same volume.
+
+## Background
+
+In Kubernetes, there are two main types of storage: local and remote.
```
Comment (ianchakeres) [inline contributors/design-proposals/local-storage-pv.md]: How about switching these around and a few other edits...
"""
In Kubernetes, there are two main types of storage: remote and local.

Remote storage is typically used with persistent volumes where the data persists beyond the lifetime of the pod.

Local storage may desirable for ephemeral use where the data persists only during the lifetime of the pod. Local storage is also desirable for distributed storage systems, such as GlusterFS and Cassandra.

The main motivations for using local storage include:

Performance:  Local SSDs can achieve higher IOPS and throughput than remote storage.

Cost: Operational costs may be reduced by leveraging existing local storage, especially in bare metal environments. Network storage can be expensive to setup and maintain, and it may not be necessary for certain applications. 

"""

File: contributors/design-proposals/local-storage-pv.md [hunk 14]
```diff
@@ -0,0 +1,632 @@
...
+
+## Feature Plan
+
+A detailed implementation plan can be found in the
+[Storage SIG planning spreadhseet](https://docs.google.com/spreadsheets/d/1t4z5DYKjX2ZDlkTpCnp18icRAQqOE85C1T1r2gqJVck/view#gid=1566770776).
+The following is a high level summary of the goals in each phase.
+
+### Phase 1
+
+* Support Pod, Deployment, and StatefulSet requesting local storage
```
Comment (ianchakeres) [inline contributors/design-proposals/local-storage-pv.md]: + for a single local volume.

File: contributors/design-proposals/local-storage-pv.md [hunk 15]
```diff
@@ -0,0 +1,632 @@
...
+* Support pre-configured, statically partitioned, filesystem-based local storage
+
+### Phase 2
+
+* Block devices and raw partitions
+* Smarter PV binding to consider local storage and pod scheduling constraints
```
Comment (ianchakeres) [inline contributors/design-proposals/local-storage-pv.md]: add: e.g. multiple local volumes per pod.

File: contributors/design-proposals/local-storage-pv.md (hunk 13 above)
Comment (msau42) [reply contributors/design-proposals/local-storage-pv.md]: done

File: contributors/design-proposals/local-storage-pv.md (hunk 14 above)
Comment (msau42) [reply contributors/design-proposals/local-storage-pv.md]: done

File: contributors/design-proposals/local-storage-pv.md (hunk 15 above)
Comment (msau42) [reply contributors/design-proposals/local-storage-pv.md]: done

File: contributors/design-proposals/local-storage-pv.md (hunk 10 above)
Comment (msau42) [reply contributors/design-proposals/local-storage-pv.md]: Yes, thanks for catching, will remove.

File: contributors/design-proposals/local-storage-pv.md (hunk 11 above)
Comment (msau42) [reply contributors/design-proposals/local-storage-pv.md]: done

File: contributors/design-proposals/local-storage-pv.md (hunk 12 above)
Comment (msau42) [reply contributors/design-proposals/local-storage-pv.md]: done

File: contributors/design-proposals/local-storage-pv.md (hunk 9 above)
Comment (msau42) [reply contributors/design-proposals/local-storage-pv.md]: Split up this section into deliverables and limitations.

File: contributors/design-proposals/local-storage-pv.md (hunk 1 above)
Comment (msau42) [reply contributors/design-proposals/local-storage-pv.md]: I updated this section to mention that it will be addressed in a future release, and added a high level overview at the bottom.

File: contributors/design-proposals/local-storage-pv.md (hunk 5 above)
Comment (msau42) [reply contributors/design-proposals/local-storage-pv.md]: Agree, I tried to reword it to make it sound scarier.

File: contributors/design-proposals/local-storage-pv.md [hunk 16]
```diff
@@ -0,0 +1,659 @@
...
+The basic design of the provisioner will have two separate handlers: one for PV deletion and
+cleanup, and the other for static PV creation.  A PersistentVolume informer will be created
+and its cache will be used by both handlers.
+
+PV deletion will operate on the Update event.  If the PV it provisioned changes to the “Released”
+state, and if the reclaim policy is Delete, then it will cleanup the volume and then delete the PV,
+removing it from the cache.
+
+PV creation does not operate on any informer events.  Instead, it periodically monitors the discovery
+directories, and will create a new PV for each path in the directory that is not in the PV cache.  It
```
Comment (hongchaodeng) [inline contributors/design-proposals/local-storage-pv.md]: What happens if no sub-dir exists?

Comment (msau42) [reply contributors/design-proposals/local-storage-pv.md]: nothing happens as we don't detect any volumes.

File: contributors/design-proposals/local-storage-pv.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (ddysher) [reply contributors/design-proposals/local-storage-pv.md]: hmm...  do you think this will become a general requirement where daemonset pod can be  restricted to only access local resources, not just for local volume use case?

**Comment by msau42** [comment]
Can folks try to review this week?  I would like to get this in, so that we can start extending it for 1.9 features.
